server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java

```diff
@@ -91,7 +92,7 @@ public class SearchExecutionContext extends QueryRewriteContext {
     private final MappingLookup mappingLookup;
     private final SimilarityService similarityService;
     private final BitsetFilterCache bitsetFilterCache;
-    private final TriFunction<MappedFieldType, String, Supplier<SearchLookup>, IndexFieldData<?>> indexFieldDataService;
+    private final QuadFunction<MappedFieldType, String, Supplier<SearchLookup>, FielddataOperation, IndexFieldData<?>> indexFieldDataLookup;
```

Comment:
I think by the time we have a generic four-input function we should probably just make this a functional interface called IndexFieldDataLookup? Alternatively we could try and combine the index name, lookup supplier and operation into a context object.

Reply:
Thinking about this more, I like the idea of a context object, as it would mean we could also put the `sourcePaths()` method there rather than extending SearchLookup further.

Reply:
I have integrated the changes from #88779 and it appears to have reduced the line count by around 400 or 25%!